packages/terra-functional-testing/lib/services/wdio-docker-service.js

```diff
@@ -0,0 +1,167 @@
+
+    logger.log('Deploying docker stack.');
+
+    const composeFilePath = path.resolve(__dirname, '../docker/docker-compose.yml');
+
+    await exec(`docker stack deploy -c ${composeFilePath} wdio`);
+
+    await this.awaitNetworkReady();
+  }
+
+  /**
+   * Removes the docker stack.
+   */
+  async removeStack() {
+    const { stdout: stackInfo } = await exec('docker stack ls | grep wdio || true');
+
+    if (stackInfo) {
+      logger.log('Removing docker stack.');
+
+      await exec('docker stack rm wdio');
+
+      // Ensure the services and network have been removed.
+      await this.awaitServiceRemoval();
+      await this.awaitNetworkRemoval();
+    }
+  }
+
+  /**
+   * Waits for a command to complete successfully.
+   * @param {string} command - The shell command to run.
+   * @param {func} callback - A callback function to accept or reject the result of the command. Must return a promise.
+   * @param {number} retries - The number of times to retry the command. Defaults to 30 retries.
+   * @param {number} interval - The timeout between commands in milliseconds. Defaults to every two seconds.
+   * @returns {Promise} - A promise that resolves when the callback accepts the command response.
+   */
+  async pollCommand(command, callback, retries = RETRY_COUNT, interval = POLL_INTERVAL) {
+    return new Promise((resolve, reject) => {
+      let retryCount = 0;
+      let pollTimeout = null;
+
+      const poll = async () => {
+        if (retryCount >= retries) {
+          clearTimeout(pollTimeout);
+          pollTimeout = null;
+          reject(Error(logger.format('Timeout. Exceeded retry count.')));
+        }
+
+        try {
+          const result = await exec(command);
+
+          await callback(result).then(() => resolve());
+        } catch (error) {
+          retryCount += 1;
+          pollTimeout = setTimeout(poll, interval);
+        }
+      };
+
+      pollTimeout = setTimeout(poll, interval);
+    });
+  }
+
+  /**
+   * Ensures the docker network has been shut down.
+   */
+  async awaitNetworkRemoval() {
+    await this.pollCommand('docker network ls | grep wdio || true', (result) => (
+      new Promise((resolve, reject) => {
+        const { stdout: networkStatus } = result;
+
+        // Reject if there is an active network returned.
+        if (networkStatus) {
+          reject();
+        } else {
+          resolve();
+        }
+      })));
+  }
+
+  /**
+   * Ensures the docker services have been shut down.
+   */
+  async awaitServiceRemoval() {
+    await this.pollCommand('docker service ls | grep wdio || true', (result) => (
+      new Promise((resolve, reject) => {
+        const { stdout: serviceStatus } = result;
+
+        // Reject if there is an active service returned.
+        if (serviceStatus) {
+          reject();
+        } else {
+          resolve();
+        }
+      })));
+  }
+
+  /**
+   * Ensures the docker network is ready.
+   */
+  async awaitNetworkReady() {
```

Comment:
Something I have been thinking about, and here may not be the correct place to do it, but we should considering polling to validate the deployed selenium grid is ready / available. There's been many times it's been down, test run and consumers test try to run and all will fail

Reply:
I think that's a good idea, it might make sense to be another service with a few responsibilities to validate the testing environment on the remote grid